gum/arch-mips/gummipsrelocator.c

```diff
@@ -0,0 +1,530 @@
+  {
+    ctx.delay_slot_insn = NULL;
+    ctx.delay_slot_detail = NULL;
+  }
+
+  switch (insn->id)
+  {
+    default:
+      rewritten = FALSE;
+      break;
+  }
+
+  if (!rewritten)
+  {
+    gum_mips_writer_put_bytes (ctx.output, insn->bytes, insn->size);
+    if (delay_slot_insn != NULL)
+    {
+      gum_mips_writer_put_bytes (ctx.output, delay_slot_insn->bytes,
+          delay_slot_insn->size);
+    }
+  }
+
+  return TRUE;
+}
+
+void
+gum_mips_relocator_write_all (GumMipsRelocator * self)
+{
+  guint count = 0;
+
+  while (gum_mips_relocator_write_one (self))
+    count++;
+
+  g_assert_cmpuint (count, >, 0);
+}
+
+gboolean
+gum_mips_relocator_eob (GumMipsRelocator * self)
+{
+  return self->eob || self->delay_slot_pending;
+}
+
+gboolean
+gum_mips_relocator_eoi (GumMipsRelocator * self)
+{
+  return self->eoi || self->delay_slot_pending;
+}
+
+gboolean
+gum_mips_relocator_can_relocate (gpointer address,
+                                 guint min_bytes,
+                                 GumRelocationScenario scenario,
+                                 guint * maximum,
+                                 mips_reg * available_scratch_reg)
+{
+  guint n = 0;
+  guint8 * buf;
+  GumMipsWriter cw;
+  GumMipsRelocator rl;
+  guint reloc_bytes;
+
+  buf = g_alloca (3 * min_bytes);
+  gum_mips_writer_init (&cw, buf);
+
+  gum_mips_relocator_init (&rl, address, &cw);
+
+  do
+  {
+    const cs_insn * insn;
+    gboolean safe_to_relocate_further;
+
+    reloc_bytes = gum_mips_relocator_read_one (&rl, &insn);
+    if (reloc_bytes == 0)
+      break;
+
+    n = reloc_bytes;
+
+    if (scenario == GUM_SCENARIO_ONLINE)
+    {
+      switch (insn->id)
+      {
+        case MIPS_INS_JAL:
+        case MIPS_INS_JALR:
+        case MIPS_INS_SYSCALL:
+          safe_to_relocate_further = FALSE;
+          break;
+        default:
+          safe_to_relocate_further = TRUE;
+          break;
+      }
+    }
+    else
+    {
+      safe_to_relocate_further = TRUE;
+    }
+
+    if (!safe_to_relocate_further)
+      break;
+  }
+  while (reloc_bytes < min_bytes || rl.delay_slot_pending);
+
+  if (!rl.eoi)
+  {
+    csh capstone;
+    cs_err err;
+    cs_insn * insn;
+    size_t count, i;
+    gboolean eoi;
+
+#ifdef G_LITTLE_ENDIAN
+    err = cs_open (CS_ARCH_MIPS, CS_MODE_MIPS32 | CS_MODE_LITTLE_ENDIAN,
+        &capstone);
+#else
+    err = cs_open (CS_ARCH_MIPS, CS_MODE_MIPS32 | CS_MODE_BIG_ENDIAN,
+        &capstone);
+#endif
+    g_assert_cmpint (err, == , CS_ERR_OK);
+    err = cs_option (capstone, CS_OPT_DETAIL, CS_OPT_ON);
+    g_assert_cmpint (err, ==, CS_ERR_OK);
+
+    count = cs_disasm (capstone, rl.input_cur, 1024, rl.input_pc, 0, &insn);
+    g_assert (insn != NULL);
+
+    eoi = FALSE;
+    for (i = 0; i != count && !eoi; i++)
+    {
+      cs_mips * d = &insn[i].detail->mips;
+
+      switch (insn[i].id)
+      {
+        case MIPS_INS_J:
+        {
+          cs_mips_op * op = &d->operands[0];
+          g_assert (op->type == MIPS_OP_IMM);
```

Comment:
Would be better to use `g_assert_cmpint` or `g_assert_cmpuint` here, so a failing assertion will show the actual value vs the one we expect.